flocker/control/test/test_persistence.py

```diff
@@ -249,3 +332,193 @@ def test_complex_keys(self):
+            u'version 4 configuration. No migration method '
+            u'upgrade_from_v3 could be found.'
+        )
+        self.assertEqual(e.message, expected_error)
+
+    def test_sequential_migrations(self):
+        """
+        A migration from one configuration version to another will
+        sequentially perform all necessary upgrades, e.g. v1 to v2 followed
+        by v2 to v3.
+        """
+        # Get a valid v2 config.
+        v2_config = migrate_configuration(1, 2, self.v1_config, StubMigration)
+        # Perform two sequential migrations to get from v1 to v3, starting
+        # with a v1 config.
+        result = migrate_configuration(1, 3, self.v1_config, StubMigration)
+        # Compare the v1 --> v3 upgrade to the direct result of the
+        # v2 --> v3 upgrade on the v2 config, Both should be identical
+        # and valid v3 configs.
+        self.assertEqual(result, StubMigration.upgrade_from_v2(v2_config))
+
+
+def manifestation_generator(random, _):
+    """
+    Hypothesis strategy generator to build a range of manifestations.
+    Generates a mapping of dataset IDs to ``Manifestation`` objects,
+    providing at random between 0 and 5 manifestations, with randomized
+    datasets and attributes.
+    """
+    manifestations = dict()
+    for count in range(0, random.randint(0, 5)):
+        dataset_id = unicode(UUID(int=random.getrandbits(128)))
+        manifestation = Manifestation(
+            primary=bool(random.getrandbits(1)),
+            dataset=Dataset(
+                dataset_id=dataset_id,
+                maximum_size=random.randint(1000000, 100000000)
+            )
+        )
+        manifestations[dataset_id] = manifestation
+    return manifestations
+
+
+ST_MANIFESTATIONS = st.basic(generate=manifestation_generator)
+UUIDS = st.basic(generate=lambda r, _: UUID(int=r.getrandbits(128)))
+ALPHABET = u"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijlmnopqrstuvwxyz"
+SIMPLE_TEXT = st.text(
+    alphabet=ALPHABET, min_size=4, max_size=10, average_size=6
+)
+IMAGES = st.builds(DockerImage, tag=SIMPLE_TEXT, repository=SIMPLE_TEXT)
+NONE_OR_INT = st.one_of(
+    st.just(None),
+    st.integers(min_value=1000000, max_value=100000000)
+)
+PORTS = st.builds(
+    Port,
+    internal_port=st.integers(min_value=1, max_value=32000),
+    external_port=st.integers(min_value=1, max_value=32000)
+)
+LINKS = st.builds(
+    Link,
+    local_port=st.integers(min_value=1, max_value=32000),
+    remote_port=st.integers(min_value=1, max_value=32000),
+    alias=SIMPLE_TEXT
+)
+APPLICATIONS = st.builds(
+    Application, name=SIMPLE_TEXT, image=IMAGES,
+    # XXX at the moment there is a problem including AttachedVolume
+    # instances in each Application, because the Manifestation in
+    # a volume must also be present on the manifestations in a Node.
+    # So far I cannot find a way to do this with Hypothesis.
+    volume=st.just(None),
+    ports=st.sets(PORTS, min_size=0, max_size=3),
```

Comment:
We discussed this a little on Zulip. Part of the point of using hypothesis is that it tests cases we would never have thought of. We should only limit the inputs if there is an explicit reason (e.g. it makes it faster; we deliberately don't handle that case; etc.). When we do add such limits, we should have a comment explaining why.